tailnet/multiagent.go

```diff
@@ -0,0 +1,132 @@
+package tailnet
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/uuid"
+
+	"cdr.dev/slog"
+)
+
+type MultiAgentConn interface {
+	UpdateSelf(node *Node) error
+	SubscribeAgent(agentID uuid.UUID) error
+	UnsubscribeAgent(agentID uuid.UUID) error
+	NextUpdate(ctx context.Context) ([]*Node, bool)
+	AgentIsLegacy(agentID uuid.UUID) bool
+	Close() error
+	IsClosed() bool
+}
+
+type MultiAgent struct {
+	mu     sync.RWMutex
+	closed bool
+
+	ID     uuid.UUID
+	Logger slog.Logger
+
+	AgentIsLegacyFunc func(agentID uuid.UUID) bool
+	OnSubscribe       func(enq Enqueueable, agent uuid.UUID) error
+	OnUnsubscribe     func(enq Enqueueable, agent uuid.UUID) error
+	OnNodeUpdate      func(id uuid.UUID, node *Node) error
+	OnRemove          func(id uuid.UUID)
+
+	updates   chan []*Node
+	closeOnce sync.Once
+	start     int64
+	lastWrite int64
+	// Client nodes normally generate a unique id for each connection so
+	// overwrites are really not an issue, but is provided for compatibility.
+	overwrites int64
+}
+
+func (m *MultiAgent) Init() *MultiAgent {
+	m.updates = make(chan []*Node, 128)
+	m.start = time.Now().Unix()
+	return m
+}
+
+func (m *MultiAgent) UniqueID() uuid.UUID {
+	return m.ID
+}
+
+func (m *MultiAgent) AgentIsLegacy(agentID uuid.UUID) bool {
+	return m.AgentIsLegacyFunc(agentID)
+}
+
+func (m *MultiAgent) UpdateSelf(node *Node) error {
+	return m.OnNodeUpdate(m.ID, node)
```

Comment:
closing the `MultiAgent` feels incomplete to me since it seems that we would still send updates, subscribes, and unsubscribes into the `core`, even after it's been closed.

Furthermore, `NextUpdate()` runs on is own, separate Context, rather than respecting the closed state of the `MultiAgent`.  This seems like a recipe for zombie goroutines.

Reply:
`NextUpdate()` does respect the closed state of `MultiAgent` since `m.updates` gets closed when the `MultiAgent` is closed.